### PR TITLE
clear history when content from disk is changed

### PR DESCRIPTION
**Pull request by aminroosta**
@johnfn Fix for #680 !
Let me know if you need something to change :-)

### CODE REVIEW

File: extension.ts [hunk 1]
```diff
@@ -114,6 +115,15 @@ export async function activate(context: vscode.ExtensionContext) {
 
   vscode.window.onDidChangeActiveTextEditor(handleActiveEditorChange, this);
 
+  vscode.workspace.onDidChangeTextDocument((event) => {
+    /* TODO: Remove setTimeout (https://github.com/Microsoft/vscode/issues/11339) */
+    setTimeout(() => {
+      if (event.document.isDirty === false) {
```
Comment (rebornix) [inline extension.ts]: I suppose you are typing too fast to check a boolean through `===` :)

File: extension.ts [hunk 2]
```diff
@@ -114,6 +115,15 @@ export async function activate(context: vscode.ExtensionContext) {
 
   vscode.window.onDidChangeActiveTextEditor(handleActiveEditorChange, this);
 
+  vscode.workspace.onDidChangeTextDocument((event) => {
+    /* TODO: Remove setTimeout (https://github.com/Microsoft/vscode/issues/11339) */
```
Comment (johnfn) [inline extension.ts]: Can you write a little documentation on why this is needed (isDirty not immediately set etc)

File: extension.ts [hunk 3]
```diff
@@ -217,6 +227,13 @@ async function handleKeyEvent(key: string): Promise<void> {
   });
 }
 
+function handleContentChangedFromDisk(document : vscode.TextDocument) : void {
+  _.filter(modeHandlerToEditorIdentity, { "fileName" : document.fileName})
```
Comment (rebornix) [inline extension.ts]: Keep the leading/trailing spaces consistent `{ "fileName" : document.fileName}`

File: extension.ts [hunk 4]
```diff
@@ -114,6 +115,19 @@ export async function activate(context: vscode.ExtensionContext) {
 
   vscode.window.onDidChangeActiveTextEditor(handleActiveEditorChange, this);
 
+  vscode.workspace.onDidChangeTextDocument((event) => {
+    /**
+     * Change from vscode eidtor should set document.isDirty to true but they initially don't!
```
Comment (johnfn) [inline extension.ts]: Fix typo eidtor => editor

File: extension.ts [hunk 5]
```diff
@@ -217,6 +231,13 @@ async function handleKeyEvent(key: string): Promise<void> {
   });
 }
 
+function handleContentChangedFromDisk(document : vscode.TextDocument) : void {
+  _.filter(modeHandlerToEditorIdentity, (modeHandler) => modeHandler.fileName === document.fileName)
```
Comment (aminroosta) [inline extension.ts]: @johnfn I used an arrow function instead, this way renaming variables using vscode won't break anything.

File: extension.ts (hunk 4 above)
Comment (aminroosta) [inline extension.ts]: @johnfn  Fixed

**Comment by aminroosta** [comment]
Any reason why `git commit --amend -m "new message"` create a new commit?!

**Comment by johnfn** [comment]
Generally speaking it shouldn't! You might get yourself into trouble if you've already pushed to a remote though.

**Comment by aminroosta** [comment]
I know :-)

File: extension.ts (hunk 5 above)
Comment (jpoon) [inline extension.ts]: Nit. You can also remove those brackets around `(modeHandler)`

**Comment by aminroosta** [comment]
@johnfn 1 test failed related to clipboard registers which i have already fixed in [implement ; and ,](https://github.com/VSCodeVim/Vim/pull/674)

File: extension.ts [hunk 6]
```diff
@@ -217,6 +231,13 @@ async function handleKeyEvent(key: string): Promise<void> {
...
+    .forEach((modeHandler) => {
```
Comment (jpoon) [inline extension.ts]: Same thing about the brackets here :)

**Comment by aminroosta** [comment]
Travis Ci is not friendly with clipboard tests, specially in mac builds.

File: extension.ts (hunk 6 above)
Comment (aminroosta) [inline extension.ts]: What about them?

**Comment by jpoon** [comment]
I just restarted the build. In any case, test failure aint your fault 👍

File: extension.ts (hunk 6 above)
Comment (jpoon) [inline extension.ts]: That you can remove the brackets around `(modeHandler)`

Comment (aminroosta) [inline extension.ts]: screen shot 2016-09-03 at 3 44 12 am
@jpoon  I saw other places are always using brackets even for a single argument, that's why i added those brackets 😄

**Comment by johnfn** [comment]
LGTM, thanks @aminroosta!

**Comment by rebornix** [comment]
This commit is amazing, now undo is trustworthy!

**Pull request merged**